docs/design/upstream-datamover.md

```diff
@@ -0,0 +1,110 @@
+
+
+## Detailed Design
+
+### Volsync Datamover removal
+
+All Volsync Datamover-related code and configuration will be removed from oadp-operator.
+
+#### DPA CRD
+
+For the `DataProtectionApplication` CRD, remove the `DataMover` (and related) structs, including `RetainPolicy` and `VolumeOptions`.
+Specifically, `DataMover` will be removed from the `Features` struct which is included in DPA spec.
+`Features` will be empty for now, but available for future tech preview features that need explicit configuration here.
+```
+// Features defines the configuration for the DPA to enable the tech preview features
+type Features struct {
+}
+```
+Other places where DataMover-related fields will be removed from the DPA:
+1. Remove `dataMoverImageFqin` from the list of valid keys for `UnsupportedOverrides`.
+1. Remove `vsmPluginImageFqin` from the list of valid keys for `UnsupportedOverrides`.
+1. Remove vsm from the list of default plugins.
+1. Remove related consts and labels.
+
+#### OADP controllers
+
+In addition, related controller actions, validation, and unit tests for Volsync data mover will be removed:
+1. controllers/validator.go: datamover-related validation
+1. controllers/datamover: the entire file
+1. controllers/datamover_test.go: the entire file
+1. controllers/velero.go: datamover-related velero configuration
+1. controllers/dpa_controller.go: datamover-related entries in `ReconcileBatch`
+1. pkg/common/common.go: datamover-related const definitions
+1. pkg/velero/server/config.go: datamover CRDs removed from restore priorities list
+
+#### Other OADP CRDs
+On the CRD side, the `VolumeSnapshotBackup` and `VolumeSnapshotRestore` CRDs will be removed from `config/crd/bases` and from `make bundle` processing.
+The new velero `DataUpload` and `DataDownload` will be added.
+
+#### OADP container images
+
+References to `velero-plugin-for-vsm` and `volume-snapshot-mover` images need to be removed from `config/manager/manager.yaml`, golang code, and `deploy/disconnected-prep.sh`.
+
+#### Velero code
+
+In addition to pulling in Velero 1.12 upstream changes into the OADP Velero fork, we will need to remove all of the Velero customizations in our velero and velero-plugin-for-csi repos that were added to support the Volsync data mover.
+
+### Velero Built-in Data Mover configuration
+
+No new DPA fields will be needed for VBDM -- to enable it, the user just needs to include the CSI plugin and enable the node agent. For 1.3, users should use the new `NodeAgentConfig` spec field for this, although the existing (deprecated in 1.3) `Restic` spec field will also work here..
+
+Note that unlike in the OADP Volsync data mover, with VBDM, to use the datamover for a specific backup, that backup must be created with the `--snapshot-move-data` flag to `velero backup create` or with `Spec.SnapshotMoveData` set to `true` on the Backup CR.
```

Comment:
This is still a tech preview feature or are we considering this to be GA?

Reply:
Having a field that is not hidden behind the "Feature" flag is what is what I am worried about. 

When we have Tech Preview, having explicit opt in is useful, if the field exists and someone starts using it they could never know that they are in an unsupported path. 

I think we have to determine this and we have to decide if we need to add something around this.

Reply:
@shawn-hurley We can't hide it behind a feature flag. It's in the velero backup spec, *not* in the DPA. Anyone installing Velero 1.12 with csi plugin and kopia or restic enabled has datamover enabled.

Reply:
The only way to *not* allow users to use datamover would be to either remove CSI or remove restic/kopia from the DPA.

Reply:
First, can we agreement on the problem, and it is a real one that we want to add to the open question PR?

Reply:
@shawn-hurley It's added to open questions. And I would agree that, in general, new features that are tech preview should be behind a feature flag, but I'm not seeing an implementation path here, since in Velero 1.12, every component needed to enable datamover backups is either 1) already available without a feature flag on the DPA (CSI plugin, restic/node-agent) or 2) not part of any OADP-specific APIs that we control (regular spec field on velero Backup). So for 1.3 and this datamover, I think the best we can do is either documentation, or we could add a validation that add a warning event if both CSI and restic/node-agent are enabled, warning that this combination will enable a tech preview feature, and that they should be aware that datamover is still tech preview.